File: crates/rattler_build_core/src/debug.rs
//! Cross-platform debug utilities for rattler-build.
//!
//! This module provides the core logic for setting up, inspecting, and running
//! debug build environments. It is used by both the CLI (`rattler-build debug`)
//! and the Python bindings (`DebugSession`).

use std::{
    path::{Path, PathBuf},
    process::Command,
};

use rattler_conda_types::{ChannelUrl, MatchSpec, ParseStrictness, PrefixRecord};

use crate::render::resolved_dependencies::RunExportsDownload;
use crate::source::patch::apply_patch_custom;
use crate::tool_configuration::Configuration;

/// Result of running a debug build script.
pub struct DebugRunResult {
    /// The exit code of the build script process.
    pub exit_code: i32,
    /// Captured standard output.
    pub stdout: String,
    /// Captured standard error.
    pub stderr: String,
}

/// Parse the directories JSON from environment, rattler-build-log.txt, or work directory.
///
/// Resolution order:
/// 1. If `work_dir` is explicitly given, use it directly.
/// 2. If `RATTLER_BUILD_DIRECTORIES` env var is set (inside a debug shell), parse it.
/// 3. Fall back to reading `rattler-build-log.txt` from the output directory.
pub fn parse_directories_info(
    work_dir: Option<PathBuf>,
    output_dir: &Path,
) -> std::io::Result<(PathBuf, Option<serde_json::Value>)> {
    if let Some(dir) = work_dir {
        return Ok((dir, None));
    }

    // Check if we're inside a debug shell with RATTLER_BUILD_DIRECTORIES set
    if let Ok(json_str) = std::env::var("RATTLER_BUILD_DIRECTORIES")
        && let Ok(json) = serde_json::from_str::<serde_json::Value>(&json_str)
    {
        let work_dir = json["work_dir"].as_str().ok_or_else(|| {
            std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "work_dir not found in RATTLER_BUILD_DIRECTORIES",
            )
        })?;
        return Ok((PathBuf::from(work_dir), Some(json)));
    }

    // Read from rattler-build-log.txt
    let log_file = output_dir.join("rattler-build-log.txt");
    if !log_file.exists() {
        eprintln!(
            "Error: Could not find rattler-build-log.txt at {}",
            log_file.display()
        );
        eprintln!("Hint: Run from inside a `rattler-build debug` shell, or specify --work-dir.");
        return Err(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            "rattler-build-log.txt not found",
        ));
    }

    let content = fs_err::read_to_string(&log_file)?;
    let last_line = content.lines().last().ok_or_else(|| {
        std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            "rattler-build-log.txt is empty",
        )
    })?;

    // Try to parse as JSON, fall back to plain path for backwards compatibility
    if let Ok(json) = serde_json::from_str::<serde_json::Value>(last_line) {
        let work_dir = json["work_dir"].as_str().ok_or_else(|| {
            std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "work_dir not found in JSON",
            )
        })?;
        Ok((PathBuf::from(work_dir), Some(json)))
    } else {
        // Old format: plain path
        Ok((PathBuf::from(last_line.trim()), None))
    }
}

/// Build Unix shell environment variable exports from directories JSON.
pub fn build_env_exports(json: &serde_json::Value) -> String {
    let mut env_exports = String::new();

    // Export the full JSON as RATTLER_BUILD_DIRECTORIES
    env_exports.push_str(&format!(
        "export RATTLER_BUILD_DIRECTORIES='{}'\n",
        serde_json::to_string(json).unwrap_or_default()
    ));

    // Export individual directories for convenience
    for (key, env_var) in &[
        ("recipe_path", "RATTLER_BUILD_RECIPE_PATH"),
        ("recipe_dir", "RATTLER_BUILD_RECIPE_DIR"),
        ("build_dir", "RATTLER_BUILD_BUILD_DIR"),
        ("output_dir", "RATTLER_BUILD_OUTPUT_DIR"),
        ("host_prefix", "RATTLER_BUILD_HOST_PREFIX"),
        ("build_prefix", "RATTLER_BUILD_BUILD_PREFIX"),
    ] {
        if let Some(val) = json[*key].as_str() {
            env_exports.push_str(&format!("export {}='{}'\n", env_var, val));
        }
    }

    // Export the path to the current rattler-build binary so that
    // subcommands like `$RATTLER_BUILD debug host-add` work even when
    // the system PATH points to a different (older) installation.
    if let Ok(exe) = std::env::current_exe()
        && let Ok(canonical) = exe.canonicalize()
    {
        env_exports.push_str(&format!(
            "export RATTLER_BUILD='{}'\nalias rattler-build='{}'\n",
            canonical.display(),
            canonical.display()
        ));
    }

    env_exports
}

/// Build Windows cmd.exe `set` commands from directories JSON.
#[cfg(windows)]
pub fn build_env_sets_windows(json: &serde_json::Value) -> String {
    let mut sets = String::new();

    // Set the full JSON as RATTLER_BUILD_DIRECTORIES
    sets.push_str(&format!(
        "set \"RATTLER_BUILD_DIRECTORIES={}\"&",
        serde_json::to_string(json).unwrap_or_default()
    ));

    for (key, env_var) in &[
        ("recipe_path", "RATTLER_BUILD_RECIPE_PATH"),
        ("recipe_dir", "RATTLER_BUILD_RECIPE_DIR"),
        ("build_dir", "RATTLER_BUILD_BUILD_DIR"),
        ("output_dir", "RATTLER_BUILD_OUTPUT_DIR"),
        ("host_prefix", "RATTLER_BUILD_HOST_PREFIX"),
        ("build_prefix", "RATTLER_BUILD_BUILD_PREFIX"),
    ] {
        if let Some(val) = json[*key].as_str() {
            sets.push_str(&format!("set \"{}={}\"&", env_var, val));
        }
    }

    if let Ok(exe) = std::env::current_exe()
        && let Ok(canonical) = exe.canonicalize()
    {
        sets.push_str(&format!("set \"RATTLER_BUILD={}\"&", canonical.display()));
    }

    sets
}

/// Print the debug shell welcome banner.
pub fn print_debug_banner(work_dir: &Path, directories_json: &Option<serde_json::Value>) {
    println!();
    println!("  rattler-build debug shell");
    println!("  ========================");
    println!();
    println!("  Work directory: {}", work_dir.display());

    if let Some(json) = &directories_json {
        if let Some(host_prefix) = json["host_prefix"].as_str() {
            println!("  Host prefix:    {}", host_prefix);
        }
        if let Some(build_prefix) = json["build_prefix"].as_str() {
            println!("  Build prefix:   {}", build_prefix);
        }
    }

    println!();
    println!("  Available commands:");
    println!("    rattler-build debug create-patch         Create a patch from your changes");
    println!("    rattler-build debug host-add <pkg>       Add packages to host env");
    println!("    rattler-build debug build-add <pkg>      Add packages to build env");
    println!();
    println!(
        "  The build environment has been sourced. Run `{}` to",
        build_script_hint()
    );
    println!("  execute the build script, or make changes and use create-patch.");
    println!();
    println!("  Exit with 'exit' or Ctrl+D.");
    println!();
}

/// Open an interactive debug shell in the build environment.
///
/// On Unix, this starts `$SHELL` with `build_env.sh` sourced.
/// On Windows, this starts `cmd.exe` with `build_env.bat` called.
pub fn open_debug_shell(
    work_dir: PathBuf,
    directories_json: Option<serde_json::Value>,
) -> std::io::Result<()> {
    validate_work_dir(&work_dir)?;
    print_debug_banner(&work_dir, &directories_json);
    open_debug_shell_platform(&work_dir, directories_json)
}

/// Run the build script and capture its output.
///
/// Returns a [`DebugRunResult`] with captured stdout and stderr.
/// On Unix, runs via `bash`. On Windows, runs via `cmd.exe`.
pub fn run_build_script(work_dir: &Path, trace: bool) -> std::io::Result<DebugRunResult> {
    validate_work_dir(work_dir)?;
    run_build_script_platform(work_dir, trace)
}

/// Run the build script with inherited stdio (for interactive CLI use).
///
/// Returns just the exit code. On Unix, runs via `bash`. On Windows, runs
/// via `cmd.exe`.
pub fn run_build_script_interactive(work_dir: &Path, trace: bool) -> std::io::Result<i32> {
    validate_work_dir(work_dir)?;
    run_build_script_interactive_platform(work_dir, trace)
}

// ---------------------------------------------------------------------------
// Platform-specific implementations
// ---------------------------------------------------------------------------

/// Hint text for the build script command (used in the banner).
#[cfg(unix)]
fn build_script_hint() -> &'static str {
    "bash -x conda_build.sh"
}

#[cfg(windows)]
fn build_script_hint() -> &'static str {
    "conda_build.bat"
}

/// Validate that a work directory exists.
fn validate_work_dir(work_dir: &Path) -> std::io::Result<()> {
    if !work_dir.exists() {
        eprintln!(
            "Error: Work directory does not exist: {}",
            work_dir.display()
        );
        return Err(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            format!("Work directory not found: {}", work_dir.display()),
        ));
    }
    Ok(())
}

/// Check that a required file exists in the work directory, returning a
/// descriptive IO error if it does not.
fn require_file(work_dir: &Path, name: &str) -> std::io::Result<PathBuf> {
    let path = work_dir.join(name);
    if !path.exists() {
        return Err(std::io::Error::new(
            std::io::ErrorKind::NotFound,
            format!("{} not found in {}", name, work_dir.display()),
        ));
    }
    Ok(path)
}

// -- Unix implementations --------------------------------------------------

#[cfg(unix)]
fn open_debug_shell_platform(
    work_dir: &Path,
    directories_json: Option<serde_json::Value>,
) -> std::io::Result<()> {
    let build_env = work_dir.join("build_env.sh");
    if !build_env.exists() {
        eprintln!("Warning: build_env.sh not found in {}", work_dir.display());
        eprintln!("The build environment may not have been set up yet.");
    }

    let shell = std::env::var("SHELL").unwrap_or_else(|_| "/bin/bash".to_string());

    let env_exports = directories_json
        .as_ref()
        .map(build_env_exports)
        .unwrap_or_default();

    let shell_script = if build_env.exists() {
        format!(
            "cd '{}' && {}source build_env.sh && exec {} -i",
            work_dir.display(),
            env_exports,
            shell
        )
    } else {
        format!(
            "cd '{}' && {}exec {} -i",
            work_dir.display(),
            env_exports,
            shell
        )
    };

    let status = Command::new(&shell).arg("-c").arg(&shell_script).status()?;

    if !status.success() {
        return Err(std::io::Error::other(format!(
            "shell exited with status: {}",
            status
        )));
    }
    Ok(())
}

#[cfg(unix)]
fn run_build_script_platform(work_dir: &Path, trace: bool) -> std::io::Result<DebugRunResult> {
    require_file(work_dir, "build_env.sh")?;
    require_file(work_dir, "conda_build.sh")?;

    let bash_flag = if trace { "-ex" } else { "-e" };
    let script = format!(
        "cd '{}' && source build_env.sh && bash {} conda_build.sh",
        work_dir.display(),
        bash_flag,
    );

    let output = Command::new("bash")
        .arg("-c")
        .arg(&script)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()?;

    Ok(DebugRunResult {
        exit_code: output.status.code().unwrap_or(1),
        stdout: String::from_utf8_lossy(&output.stdout).into_owned(),
        stderr: String::from_utf8_lossy(&output.stderr).into_owned(),
    })
}

#[cfg(unix)]
fn run_build_script_interactive_platform(work_dir: &Path, trace: bool) -> std::io::Result<i32> {
    require_file(work_dir, "build_env.sh")?;
    require_file(work_dir, "conda_build.sh")?;

    let bash_flag = if trace { "-ex" } else { "-e" };
    let script = format!(
        "cd '{}' && source build_env.sh && bash {} conda_build.sh",
        work_dir.display(),
        bash_flag,
    );

    let status = Command::new("bash").arg("-c").arg(&script).status()?;
    Ok(status.code().unwrap_or(1))
}

// -- Windows implementations -----------------------------------------------

#[cfg(windows)]
fn open_debug_shell_platform(
    work_dir: &Path,
    directories_json: Option<serde_json::Value>,
) -> std::io::Result<()> {
    let build_env = work_dir.join("build_env.bat");
    if !build_env.exists() {
        eprintln!("Warning: build_env.bat not found in {}", work_dir.display());
        eprintln!("The build environment may not have been set up yet.");
    }

    let comspec = std::env::var("COMSPEC").unwrap_or_else(|_| "cmd.exe".to_string());

    let env_sets = directories_json
        .as_ref()
        .map(build_env_sets_windows)
        .unwrap_or_default();

    let init_cmd = if build_env.exists() {
        format!(
            "cd /d \"{}\" && {}call build_env.bat",
            work_dir.display(),
            env_sets
        )
    } else {
        format!("cd /d \"{}\" && {}", work_dir.display(), env_sets)
    };

    let status = Command::new(&comspec)
        .arg("/d")
        .arg("/k")
        .arg(&init_cmd)
        .status()?;

    if !status.success() {
        return Err(std::io::Error::other(format!(
            "shell exited with status: {}",
            status
        )));
    }
    Ok(())
}

#[cfg(windows)]
fn run_build_script_platform(work_dir: &Path, trace: bool) -> std::io::Result<DebugRunResult> {
    require_file(work_dir, "conda_build.bat")?;

    let comspec = std::env::var("COMSPEC").unwrap_or_else(|_| "cmd.exe".to_string());

    let output = Command::new(&comspec)
        .arg("/d")
        .arg("/c")
        .arg("conda_build.bat")
        .current_dir(work_dir)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .output()?;

    Ok(DebugRunResult {
        exit_code: output.status.code().unwrap_or(1),
        stdout: String::from_utf8_lossy(&output.stdout).into_owned(),
        stderr: String::from_utf8_lossy(&output.stderr).into_owned(),
    })
}

#[cfg(windows)]
fn run_build_script_interactive_platform(work_dir: &Path, _trace: bool) -> std::io::Result<i32> {
    require_file(work_dir, "conda_build.bat")?;

    let comspec = std::env::var("COMSPEC").unwrap_or_else(|_| "cmd.exe".to_string());

    let status = Command::new(&comspec)
        .arg("/d")
        .arg("/c")
        .arg("conda_build.bat")
        .current_dir(work_dir)
        .status()?;

    Ok(status.code().unwrap_or(1))
}

// ---------------------------------------------------------------------------
// Environment modification
// ---------------------------------------------------------------------------

/// Add packages to a conda prefix, preserving existing packages.
///
/// Reads installed packages from `prefix`, locks them at their current
/// version, merges in `specs`, resolves, and installs only what changed.
pub async fn add_packages_to_prefix(
    env_name: &str,
    prefix: &Path,
    specs: &[String],
    channels: &[ChannelUrl],
    tool_config: &Configuration,
) -> miette::Result<()> {
    use miette::IntoDiagnostic;

    let new_specs: Vec<MatchSpec> = specs
        .iter()
        .map(|s| MatchSpec::from_str(s, ParseStrictness::Lenient))
        .collect::<Result<Vec<_>, _>>()
        .into_diagnostic()?;

    let existing_records = PrefixRecord::collect_from_prefix(prefix).into_diagnostic()?;

    let mut all_specs: Vec<MatchSpec> = existing_records
        .iter()
        .map(|r: &PrefixRecord| {
            MatchSpec::from_str(
                &format!(
                    "{}={}={}",
                    r.repodata_record.package_record.name.as_normalized(),
                    r.repodata_record.package_record.version,
                    r.repodata_record.package_record.build,
                ),
                ParseStrictness::Lenient,
            )
            .expect("existing package record should parse as MatchSpec")
        })
        .collect();
    all_specs.extend(new_specs);

    let platform_with_vp = crate::metadata::PlatformWithVirtualPackages::detect(
        &rattler_virtual_packages::VirtualPackageOverrides::default(),
    )
    .into_diagnostic()?;

    tracing::info!(
        "\nAdding {} new spec(s) to {} environment ({} existing packages)",
        specs.len(),
        env_name,
        existing_records.len(),
    );

    crate::render::solver::create_environment(
        env_name,
        &all_specs,
        &platform_with_vp,
        prefix,
        channels,
        tool_config,
        rattler_solve::ChannelPriority::Strict,
        rattler_solve::SolveStrategy::default(),
        None,
    )
    .await?;

    tracing::info!("\nSuccessfully added packages to {} environment.", env_name);

    Ok(())
}

// ---------------------------------------------------------------------------
// Output extension
// ---------------------------------------------------------------------------

impl crate::metadata::Output {
    /// Set up a debug environment without running the build.
    ///
    /// This consolidates the 5-step setup sequence:
    /// 1. Recreate directories
    /// 2. Fetch sources
    /// 3. Resolve dependencies
    /// 4. Install environments
    /// 5. Create build script
    pub async fn setup_debug_environment(
        self,
        tool_config: &Configuration,
    ) -> miette::Result<crate::metadata::Output> {
        use miette::IntoDiagnostic;

        self.build_configuration
            .directories
            .recreate_directories()
            .into_diagnostic()?;

        let output: crate::metadata::Output = self
            .fetch_sources(tool_config, apply_patch_custom)
            .await
            .into_diagnostic()?;

        let output: crate::metadata::Output = output
            .resolve_dependencies(tool_config, RunExportsDownload::DownloadMissing)
            .await
            .into_diagnostic()?;

        output
            .install_environments(tool_config)
            .await
            .into_diagnostic()?;

        output.create_build_script().await.into_diagnostic()?;

        Ok(output)
    }
}

File: py-rattler-build/src/rattler_build/debug.py
"""Debug session API for interactive build debugging.

The :class:`DebugSession` sets up a full build environment (resolves
dependencies, fetches sources, installs environments, creates build
script) **without** running the build. You can then iteratively run and
re-run the build script, inspecting output each time.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

from rattler_build._rattler_build import debug as _debug

if TYPE_CHECKING:
    from rattler_build.progress import ProgressCallback
    from rattler_build.render import RenderedVariant
    from rattler_build.tool_config import ToolConfiguration


@dataclass(frozen=True)
class ScriptResult:
    """Result of running a build script.

    Non-zero ``exit_code`` is **not** an exception — it is expected during
    debugging.  Inspect ``stdout`` and ``stderr`` to diagnose the failure.
    """

    exit_code: int
    """Process exit code (0 = success)."""

    stdout: str
    """Captured standard output."""

    stderr: str
    """Captured standard error."""

    @property
    def success(self) -> bool:
        """Whether the script exited successfully."""
        return self.exit_code == 0


@dataclass(frozen=True)
class DebugPaths:
    """All paths associated with a debug session."""

    work_dir: Path
    """Work directory where the build script and sources live."""

    host_prefix: Path
    """Host prefix (where host dependencies are installed)."""

    build_prefix: Path
    """Build prefix (where build dependencies are installed)."""

    build_script: Path
    """Path to the build script (``conda_build.sh`` / ``conda_build.bat``)."""

    build_env_script: Path
    """Path to the environment script (``build_env.sh`` / ``build_env.bat``)."""

    build_dir: Path
    """Build directory (parent of work, host_env, build_env)."""

    output_dir: Path
    """Output directory where packages are written."""

    recipe_dir: Path
    """Recipe directory."""


class DebugSession:
    """Interactive debugging session for a conda recipe.

    Use :meth:`create` to set up the environment. Then call :meth:`run_script`
    to execute the build, inspect results, modify files, and re-run.

    Example::

        session = DebugSession.create(variant, channels=["conda-forge"])
        result = session.run_script(trace=True)
        print(result.stdout)
    """

    def __init__(self, inner: _debug.DebugSession) -> None:
        self._inner = inner

    @classmethod
    def create(
        cls,
        variant: RenderedVariant,
        *,
        tool_config: ToolConfiguration | None = None,
        output_dir: str | Path | None = None,
        channels: list[str] | None = None,
        no_build_id: bool = True,
        progress_callback: ProgressCallback | None = None,
    ) -> DebugSession:
        """Create a debug session from a rendered variant.

        This resolves dependencies, fetches sources, installs environments,
        and creates the build script — all without running the actual build.

        Parameters
        ----------
        variant:
            A rendered variant from ``recipe.render()``.
        tool_config:
            Optional tool configuration. Uses defaults if not provided.
        output_dir:
            Where to write build artifacts. A temporary directory is used if
            not specified.
        channels:
            Conda channels to use. Defaults to ``["conda-forge"]``.
        no_build_id:
            If ``True`` (default), omit the build ID from directory paths,
            giving stable paths across iterations.
        progress_callback:
            Optional callback for progress events during setup.

        Returns
        -------
        DebugSession
            A session with the environment fully set up and ready to run.
        """
        output_dir_path = Path(output_dir) if output_dir is not None else None
        tc = tool_config._inner if tool_config is not None else None
        inner = _debug.create_debug_session_py(
            rendered_variant=variant._inner,
            tool_config=tc,
            output_dir=output_dir_path,
            channels=channels,
            no_build_id=no_build_id,
            progress_callback=progress_callback,
            v3=variant._v3,
        )
        return cls(inner)

    @property
    def paths(self) -> DebugPaths:
        """All paths associated with this debug session."""
        return DebugPaths(
            work_dir=Path(self._inner.work_dir),
            host_prefix=Path(self._inner.host_prefix),
            build_prefix=Path(self._inner.build_prefix),
            build_script=Path(self._inner.build_script),
            build_env_script=Path(self._inner.build_env_script),
            build_dir=Path(self._inner.build_dir),
            output_dir=Path(self._inner.output_dir),
            recipe_dir=Path(self._inner.recipe_dir),
        )

    # Convenience accessors (match the old tutorial API)

    @property
    def work_dir(self) -> Path:
        """Work directory where the build script and sources live."""
        return Path(self._inner.work_dir)

    @property
    def host_prefix(self) -> Path:
        """Host prefix (where host dependencies are installed)."""
        return Path(self._inner.host_prefix)

    @property
    def build_prefix(self) -> Path:
        """Build prefix (where build dependencies are installed)."""
        return Path(self._inner.build_prefix)

    @property
    def build_script(self) -> Path:
        """Path to the build script."""
        return Path(self._inner.build_script)

    @property
    def output_dir(self) -> Path:
        """Output directory where packages are written."""
        return Path(self._inner.output_dir)

    @property
    def setup_log(self) -> list[str]:
        """Log messages captured during environment setup."""
        return self._inner.setup_log

    @property
    def log(self) -> list[str]:
        """Alias for :attr:`setup_log` (backwards compatibility)."""
        return self._inner.setup_log

    def run_script(self, *, trace: bool = False) -> ScriptResult:
        """Run the build script and capture output.

        Parameters
        ----------
        trace:
            If ``True``, run bash with ``-x`` flag to trace each command.

        Returns
        -------
        ScriptResult
            The exit code, stdout, and stderr from the build script.
        """
        exit_code, stdout, stderr = self._inner.run_script(trace=trace)
        return ScriptResult(exit_code=exit_code, stdout=stdout, stderr=stderr)

    # Alias for backwards compatibility with old tutorial
    run = run_script

    def add_packages(
        self,
        specs: list[str],
        *,
        environment: str = "host",
        channels: list[str] | None = None,
    ) -> list[str]:
        """Add packages to the host or build environment.

        Parameters
        ----------
        specs:
            Conda match specs (e.g. ``["numpy", "pandas>=2"]``).
        environment:
            ``"host"`` or ``"build"``.
        channels:
            Override channels for this operation.

        Returns
        -------
        list[str]
            The specs that were requested (confirmation).
        """
        return self._inner.add_packages(specs, environment=environment, channels=channels)

    def create_patch(
        self,
        name: str = "changes",
        *,
        output_dir: str | Path | None = None,
        overwrite: bool = False,
        exclude: list[str] | None = None,
        add: list[str] | None = None,
        include: list[str] | None = None,
        dry_run: bool = False,
    ) -> str:
        """Create a patch from changes in the work directory.

        Parameters
        ----------
        name:
            Patch file name (without extension).
        output_dir:
            Where to write the patch file. Defaults to the recipe directory.
        overwrite:
            Overwrite existing patch file.
        exclude:
            Glob patterns to exclude from the patch.
        add:
            Glob patterns for untracked files to include.
        include:
            Glob patterns to restrict the patch to.
        dry_run:
            If ``True``, print the patch without writing it.

        Returns
        -------
        str
            Empty string on success.
        """
        output_dir_path = Path(output_dir) if output_dir is not None else None
        return self._inner.create_patch(
            name=name,
            output_dir=output_dir_path,
            overwrite=overwrite,
            exclude=exclude,
            add=add,
            include=include,
            dry_run=dry_run,
        )

    def read_build_script(self) -> str:
        """Read and return the build script contents."""
        return self._inner.read_build_script()

    def __repr__(self) -> str:
        return f"DebugSession(work_dir='{self.work_dir}')"
